drivers/usb/device/usb_dc_stm32.c

```diff
@@ -1044,7 +1045,36 @@ int usb_dc_detach(void)
 
 int usb_dc_reset(void)
 {
-	LOG_ERR("Not implemented");
+	uint32_t cnt = 0U;
+
+	/* Wait for AHB master idle state. */
+	while (!(USB_CORE_REG->GRSTCTL & USB_OTG_GRSTCTL_AHBIDL)) {
```

Comment:
This is only valid on series that have a USB OTG (`st,stm32-otgfs`), which excludes F0/F3/G0/G4/L0/(some)L4.
So either it's possible to abstract for other kind of USB HW blocks, either you need to restrict this code under this compatible for now.